File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.swift
import Foundation

func solution(_ fees:[Int], _ records:[String]) -> [Int] {
    // records는 최대 1000개
    
    // O(n)
    let times: [String: Int] = calculateTime(fees, records)
    
    // O(nlogn) -> 1000대
    let cars: [String] = times.keys.sorted() // 자동차 번호판 문자열 배열
    var totalFee: [Int] = []
    
    // O(n)
    for car in cars {
        totalFee.append(caculateFee(fees, times[car]!))
    }
    
    return totalFee // 차량번호 작은 순으로 요금 정렬
}

// 주차 요금 계산
// 입차, 출차 기록

// 기본시간 / 기본요금 / 단위 시간 / 단위 요금

// fee [기본시간, 기본요금, 단위시간, 단위요금]
// record "HH:MM 0000 IN"
// record "HH:MM 0000 OUT"

// 1. 차량 별 누적시간 -> Dictionary에 저장, 누적시간["차량번호"]
// 2. 딕셔너리.keys -> 배열, 차량번호별 정렬 > 차[차량인덱스] = 차량번호
// 2. 차량 별 요금 -> 요금[차량인덱스] = 요금
func calculateTime(_ fees:[Int], _ records: [String]) -> [String: Int] {
    var totalTime: [String: Int] = [:]
    var curRecords: [String: Int] = [:]
    
    for record in records {
        let TNIO = record.split(separator: " ").map { String($0) }
        let (T,N,IO) = (TNIO[0], TNIO[1], TNIO[2])
        
        let min = Int(String(T.prefix(2)))! * 60 + Int(String(T.suffix(2)))!
        
        if IO == "IN" { // 차량 진입
            curRecords[N] = min // 시간 기록
        } else { // 차량 출차
            totalTime[N, default: 0] += min - curRecords[N]! // 누적 시간 기록
            curRecords[N] = nil // 시간 기록 초기화
        }
    }
        
    // 남은 23:59 이후 기록
    let lastTime = 23 * 60 + 59
    for (carNum, inTime) in curRecords {
        totalTime[carNum, default: 0] += lastTime - inTime
    }
    
    return totalTime
}

// 1. 23:59에 출차된 것으로 간주
// 2. 00:00 ~ 23:59를 일괄 정산
// 3. 누적 < 기본시간 -> 기본 요금
// 4. 누적 >= 기본시간 -> 기본요금 + 올림(초과 시간 / 단위 시간) * 단위 요금
func caculateFee(_ fees:[Int], _ time: Int) -> Int {
    let baseT = fees[0] // 기본시간 1~1439
    let baseF = fees[1] // 기본요금 0~10만
    let unitT = fees[2] // 단위시간 1~1439
    let unitF = fees[3] // 단위요금 1~1만
    
    if time < baseT { 
        return baseF 
    }
    
    // 기본시간 초과
    return baseF + ((time - baseT + unitT - 1) / unitT) * unitF
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
# 주차요금 - 입차 출차

# 출차없음 > 23:59에 출차

# 기본 시간 이하 > 기본요금
# 기본 시간 초과 > 기본 요금 + 단위요금 * ceil(초과시간 / 단위시간)

# 차량번호가 작은 자동차부터 청구할 주차요금 배열 리턴

# base_t, base_f, unit_t, unit_f

# [시각 차량번호 내역]
# [HH:MM 0000 IN/OUT]
# 시각을 기준으로 오름차순

# 해당 차량 > { 0000: [IN_t, OUT_t, ] } dict로 저장
# 분 단위로 변경

# 1. records를 가공 > dict에 저장

import math
from collections import deque

def solution(fees, records):
    
    base_t, base_f, unit_t, unit_f = fees
    
    IN_OUT = {}
    
    # 데이터 가공 { 0000: [IN_t, OUT_t, IN_t, ...] }
    for record in records:
        HHMM, car_num, state = record.split()
        IN_OUT.setdefault(car_num, deque()).append(HHMM_to_minute(HHMM))
    
    
    parking_fees = [] #  [ (차량번호, 금액) ]
    
    for car_num in IN_OUT.keys():
        
        if len(IN_OUT[car_num]) % 2 == 1:
            IN_OUT[car_num].append(HHMM_to_minute("23:59"))
        
        parking_t = 0
        
        while IN_OUT[car_num]:
            st_t, end_t = IN_OUT[car_num].popleft(), IN_OUT[car_num].popleft()
            parking_t += end_t - st_t
        
        if parking_t <= base_t:
            total_f = base_f
        else:
            total_f = base_f + unit_f * math.ceil((parking_t - base_t) / unit_t) 
        
        parking_fees.append((car_num, total_f))
    
    parking_fees.sort(key=lambda x: x[0])
    
    return list(map(lambda x: x[1], parking_fees))

def HHMM_to_minute(HHMM):
    H, M = map(int, HHMM.split(":"))
    return H * 60 + M
    
    
    
    
    